File: Freedom/微信/微信/ChatViewController/ChatEmojiKeyboard/TLEmojiBaseCell.m
//  TLEmojiBaseCell.m
//  Freedom
// Created by Super
#import "TLEmojiBaseCell.h"
#import "NSFileManager+expanded.h"
@implementation TLEmoji
+ (NSDictionary *)replacedKeyFromPropertyName{
    return @{
             @"emojiID" : @"pId",
             @"emojiURL" : @"Url",
             @"emojiName" : @"credentialName",
             @"emojiPath" : @"imageFile",
             @"size" : @"size",
             };
}
- (NSString *)emojiPath{
    if (_emojiPath == nil) {
        NSString *groupPath = [NSFileManager pathExpressionForGroupID:self.groupID];
        _emojiPath = [NSString stringWithFormat:@"%@%@", groupPath, self.emojiID];
    }
    return _emojiPath;
}
@end
@implementation TLEmojiGroup
+ (NSDictionary *)replacedKeyFromPropertyName{
    return @{
             @"groupID" : @"eId",
             @"groupIconURL" : @"coverUrl",
             @"groupName" : @"eName",
             @"groupInfo" : @"memo",
             @"groupDetailInfo" : @"memo1",
             @"count" : @"picCount",
             @"bannerID" : @"aId",
             @"bannerURL" : @"URL",
             };
}
- (id)init{
    if (self = [super init]) {
        [self setType:TLEmojiTypeImageWithTitle];
    }
    return self;
}
- (void)setType:(TLEmojiType)type{
    _type = type;
    switch (type) {
        case TLEmojiTypeOther:
            return;
        case TLEmojiTypeFace:
        case TLEmojiTypeEmoji:
            self.rowNumber = 3;
            self.colNumber = 7;
            break;
        case TLEmojiTypeImage:
        case TLEmojiTypeFavorite:
        case TLEmojiTypeImageWithTitle:
            self.rowNumber = 2;
            self.colNumber = 4;
            break;
        default:
            break;
    }
    self.pageItemCount = self.rowNumber * self.colNumber;
    self.pageNumber = self.count / self.pageItemCount + (self.count % self.pageItemCount == 0 ? 0 : 1);
}
- (void)setData:(NSMutableArray *)data{
    _data = data;
    self.count = data.count;
    self.pageItemCount = self.rowNumber * self.colNumber;
    self.pageNumber = self.count / self.pageItemCount + (self.count % self.pageItemCount == 0 ? 0 : 1);
}
- (id)objectAtIndex:(NSUInteger)index{
    return [self.data objectAtIndex:index];
}
- (NSString *)path{
    if (_path == nil && self.groupID != nil) {
        _path = [NSFileManager pathExpressionForGroupID:self.groupID];
    }
    return _path;
}
- (NSString *)groupIconPath{
    if (_groupIconPath == nil && self.path != nil) {
        _groupIconPath = [NSString stringWithFormat:@"%@icon_%@", self.path, self.groupID];
    }
    return _groupIconPath;
}
- (NSString *)pictureURL{
    return self.bannerURL;
}
@end
@interface TLEmojiFaceItemCell ()
@property (nonatomic, strong) UIImageView *imageView;
@end
@implementation TLEmojiFaceItemCell
- (id) initWithFrame:(CGRect)frame{
    if (self = [super initWithFrame:frame]) {
        [self.contentView addSubview:self.imageView];
        [self p_addMasonry];
    }
    return self;
}
- (void)setEmojiItem:(TLEmoji *)emojiItem{
    [super setEmojiItem:emojiItem];
    [self.imageView setImage:emojiItem.emojiName == nil ? nil : [UIImage imageNamed:emojiItem.emojiName]];
}
#pragma mark - Private Methods -
- (void)p_addMasonry{
    [self.imageView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.edges.mas_equalTo(self);
    }];
}
#pragma mark - Getter -
- (UIImageView *)imageView{
    if (_imageView == nil) {
        _imageView = [[UIImageView alloc] init];
    }
    return _imageView;
}
@end
@interface TLEmojiImageItemCell ()
@property (nonatomic, strong) UIImageView *imageView;
@end
@implementation TLEmojiImageItemCell
- (id) initWithFrame:(CGRect)frame{
    if (self = [super initWithFrame:frame]) {
        [self.contentView addSubview:self.imageView];
        [self setHighlightImage:[UIImage imageNamed:@"emoji_hl_background"]];
        [self p_addMasonry];
    }
    return self;
}
- (void)setEmojiItem:(TLEmoji *)emojiItem{
    [super setEmojiItem:emojiItem];
    [self.imageView setImage:emojiItem.emojiPath == nil ? nil : [UIImage imageNamed:emojiItem.emojiPath]];
}
- (void)p_addMasonry{
    [self.imageView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.edges.mas_equalTo(self.contentView).mas_offset(UIEdgeInsetsMake(2, 2, 2, 2));
    }];
}
- (UIImageView *)imageView{
    if (_imageView == nil) {
        _imageView = [[UIImageView alloc] init];
    }
    return _imageView;
}
@end
@interface TLEmojiImageTitleItemCell ()
@property (nonatomic, strong) UIImageView *imageView;
@property (nonatomic, strong) UILabel *label;
@end
@implementation TLEmojiImageTitleItemCell
- (id) initWithFrame:(CGRect)frame{
    if (self = [super initWithFrame:frame]) {
        [self setHighlightImage:[UIImage imageNamed:@"emoji_hl_background"]];
        [self.contentView addSubview:self.imageView];
        [self.contentView addSubview:self.label];
        [self p_addMasonry];
    }
    return self;
}
- (void)setEmojiItem:(TLEmoji *)emojiItem{
    [super setEmojiItem:emojiItem];
    [self.imageView setImage:emojiItem.emojiPath == nil ? nil : [UIImage imageNamed:emojiItem.emojiPath]];
    [self.label setText:emojiItem.emojiName];
}
#pragma mark - Private Methods -
- (void)p_addMasonry{
    [self.bgView mas_remakeConstraints:^(MASConstraintMaker *make) {
        make.left.and.right.and.top.mas_equalTo(self.contentView);
        make.height.mas_equalTo(self.bgView.mas_width);
    }];
    [self.imageView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.left.and.top.mas_equalTo(self.contentView).mas_offset(3);
        make.right.mas_equalTo(self.contentView).mas_offset(-3);
        make.height.mas_equalTo(self.imageView.mas_width);
    }];
    [self.label mas_makeConstraints:^(MASConstraintMaker *make) {
        make.left.and.right.and.bottom.mas_equalTo(self.contentView);
    }];
}
#pragma mark - Getter -
- (UIImageView *)imageView{
    if (_imageView == nil) {
        _imageView = [[UIImageView alloc] init];
    }
    return _imageView;
}
- (UILabel *)label{
    if (_label == nil) {
        _label = [[UILabel alloc] init];
        [_label setFont:[UIFont systemFontOfSize:12.0f]];
        [_label setTextColor:[UIColor grayColor]];
        [_label setTextAlignment:NSTextAlignmentCenter];
    }
    return _label;
}
@end
@interface TLEmojiItemCell ()
@property (nonatomic, strong) UILabel *label;
@end
@implementation TLEmojiItemCell
- (id) initWithFrame:(CGRect)frame{
    if (self = [super initWithFrame:frame]) {
        [self.contentView addSubview:self.label];
        [self p_addMasonry];
    }
    return self;
}
- (void)setEmojiItem:(TLEmoji *)emojiItem{
    [super setEmojiItem:emojiItem];
    [self.label setText:emojiItem.emojiName];
}
#pragma mark - Private Methods -
- (void)p_addMasonry{
    [self.label mas_makeConstraints:^(MASConstraintMaker *make) {
        make.center.mas_equalTo(self);
    }];
}
#pragma mark - Getter -
- (UILabel *)label{
    if (_label == nil) {
        _label = [[UILabel alloc] init];
        [_label setFont:[UIFont systemFontOfSize:25.0f]];
    }
    return _label;
}
@end
@implementation TLEmojiBaseCell
- (id)initWithFrame:(CGRect)frame{
    if (self = [super initWithFrame:frame]) {
        [self.contentView addSubview:self.bgView];
        [self.bgView mas_makeConstraints:^(MASConstraintMaker *make) {
            make.edges.mas_equalTo(self.contentView);
        }];
    }
    return self;
}
- (void)setShowHighlightImage:(BOOL)showHighlightImage{
    if (showHighlightImage) {
        [self.bgView setImage:self.highlightImage];
    }else{
        [self.bgView setImage:nil];
    }
}
#pragma mark - Getter -
- (UIImageView *)bgView{
    if (_bgView == nil) {
        _bgView = [[UIImageView alloc] init];
        [_bgView.layer setMasksToBounds:YES];
        [_bgView.layer setCornerRadius:5.0f];
    }
    return _bgView;
}
@end
@interface TLEmojiGroupCell ()
@property (nonatomic, strong) UIImageView *groupIconView;
@end
@implementation TLEmojiGroupCell
- (id) initWithFrame:(CGRect)frame{
    if (self = [super initWithFrame:frame]) {
        [self setBackgroundColor:[UIColor clearColor]];
        UIView *selectedView = [[UIView alloc] init];
        [selectedView setBackgroundColor:RGBACOLOR(245.0, 245.0, 247.0, 1.0)];
        [self setSelectedBackgroundView:selectedView];
        
        [self.contentView addSubview:self.groupIconView];
        [self p_addMasonry];
    }
    return self;
}
- (void)setEmojiGroup:(TLEmojiGroup *)emojiGroup{
    _emojiGroup = emojiGroup;
    [self.groupIconView setImage:[UIImage imageNamed:emojiGroup.groupIconPath]];
}
#pragma mark - Private Methods -
- (void) p_addMasonry{
    [self.groupIconView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.center.mas_equalTo(self.contentView);
        make.width.and.height.mas_lessThanOrEqualTo(30);
    }];
}
- (void)drawRect:(CGRect)rect{
    [super drawRect:rect];
    CGContextRef context = UIGraphicsGetCurrentContext();
    CGContextSetLineWidth(context, 0.5);
    CGContextSetStrokeColorWithColor(context, colorGrayLine.CGColor);
    CGContextBeginPath(context);
    CGContextMoveToPoint(context, self.frame.size.width - 0.5, 5);
    CGContextAddLineToPoint(context, self.frame.size.width - 0.5, self.frameHeight - 5);
    CGContextStrokePath(context);
}
#pragma mark - Getter -
- (UIImageView *)groupIconView{
    if (_groupIconView == nil) {
        _groupIconView = [[UIImageView alloc] init];
    }
    return _groupIconView;
}
@end
#define     WIDTH_EMOJIGROUP_CELL       46
#define     WIDTH_SENDBUTTON            60
@interface TLEmojiGroupControl () <UICollectionViewDataSource, UICollectionViewDelegate, UICollectionViewDelegateFlowLayout>
@property (nonatomic, strong) NSIndexPath *curIndexPath;
@property (nonatomic, strong) UIButton *addButton;
@property (nonatomic, strong) UICollectionView *collectionView;
@property (nonatomic, strong) UIButton *sendButton;
@end
@implementation TLEmojiGroupControl
- (id) init{
    if (self = [super init]) {
        [self setBackgroundColor:[UIColor whiteColor]];
        [self addSubview:self.addButton];
        [self addSubview:self.collectionView];
        [self addSubview:self.sendButton];
        [self p_addMasonry];
        
        [self.collectionView registerClass:[TLEmojiGroupCell class] forCellWithReuseIdentifier:@"TLEmojiGroupCell"];
    }
    return self;
}
- (void)setSendButtonStatus:(TLGroupControlSendButtonStatus)sendButtonStatus{
    if (_sendButtonStatus != sendButtonStatus) {
        if (_sendButtonStatus == TLGroupControlSendButtonStatusNone) {
            [UIView animateWithDuration:BORDER_WIDTH_1PX animations:^{
                [self.sendButton mas_updateConstraints:^(MASConstraintMaker *make) {
                    make.right.mas_equalTo(self);
                }];
                [self layoutIfNeeded];
            }];
        }
        
        _sendButtonStatus = sendButtonStatus;
        if (sendButtonStatus == TLGroupControlSendButtonStatusNone) {
            [UIView animateWithDuration:BORDER_WIDTH_1PX animations:^{
                [self.sendButton mas_updateConstraints:^(MASConstraintMaker *make) {
                    make.right.mas_equalTo(self).mas_offset(WIDTH_SENDBUTTON);
                }];
                [self layoutIfNeeded];
            }];
        }else if (sendButtonStatus == TLGroupControlSendButtonStatusBlue) {
            [_sendButton setBackgroundImage:[UIImage imageNamed:@"emojiKB_sendBtn_blue"] forState:UIControlStateNormal];
            [_sendButton setBackgroundImage:[UIImage imageNamed:@"emojiKB_sendBtn_blueHL"] forState:UIControlStateHighlighted];
            [_sendButton setTitleColor:[UIColor whiteColor] forState:UIControlStateNormal];
        }else if (sendButtonStatus == TLGroupControlSendButtonStatusGray) {
            [_sendButton setBackgroundImage:[UIImage imageNamed:@"emojiKB_sendBtn_gray"] forState:UIControlStateNormal];
            [_sendButton setBackgroundImage:[UIImage imageNamed:@"emojiKB_sendBtn_gray"] forState:UIControlStateHighlighted];
            [_sendButton setTitleColor:[UIColor grayColor] forState:UIControlStateNormal];
        }
    }
}
- (void)setEmojiGroupData:(NSMutableArray *)emojiGroupData{
    if (_emojiGroupData == emojiGroupData || [_emojiGroupData isEqualToArray:emojiGroupData]) {
        return;
    }
    _emojiGroupData = emojiGroupData;
    [self.collectionView reloadData];
    if (emojiGroupData && emojiGroupData.count > 0) {
        [self setCurIndexPath:[NSIndexPath indexPathForRow:0 inSection:0]];
    }
}
- (void)setCurIndexPath:(NSIndexPath *)curIndexPath{
    [self.collectionView selectItemAtIndexPath:curIndexPath animated:NO scrollPosition:UICollectionViewScrollPositionNone];
    if (_curIndexPath && _curIndexPath.section == curIndexPath.section && _curIndexPath.row == curIndexPath.row) {
        return;
    }
    _curIndexPath = curIndexPath;
    if (_delegate && [_delegate respondsToSelector:@selector(emojiGroupControl:didSelectedGroup:)]) {
        TLEmojiGroup *group = [self.emojiGroupData[curIndexPath.section] objectAtIndex:curIndexPath.row];
        [_delegate emojiGroupControl:self didSelectedGroup:group];
    }
}
#pragma mark - Delegate -
//MARK: UICollectionViewDataSource
- (NSInteger)numberOfSectionsInCollectionView:(UICollectionView *)collectionView{
    return self.emojiGroupData.count;
}
- (NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section{
    return [(NSArray*)self.emojiGroupData[section] count];
}
- (UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath{
    TLEmojiGroupCell *cell = [collectionView dequeueReusableCellWithReuseIdentifier:@"TLEmojiGroupCell" forIndexPath:indexPath];
    TLEmojiGroup *group = [self.emojiGroupData[indexPath.section] objectAtIndex:indexPath.row];
    [cell setEmojiGroup:group];
    return cell;
}
- (CGSize)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout *)collectionViewLayout sizeForItemAtIndexPath:(NSIndexPath *)indexPath{
    return CGSizeMake(WIDTH_EMOJIGROUP_CELL, self.frameHeight);
}
- (CGSize)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout *)collectionViewLayout referenceSizeForFooterInSection:(NSInteger)section{
    if (section == self.emojiGroupData.count - 1) {
        return CGSizeMake(WIDTH_EMOJIGROUP_CELL * 2, self.frameHeight);
    }
    return CGSizeZero;
}
//MARK: UICollectionViewDelegate
- (void)collectionView:(UICollectionView *)collectionView didSelectItemAtIndexPath:(NSIndexPath *)indexPath{
    TLEmojiGroup *group = [self.emojiGroupData[indexPath.section] objectAtIndex:indexPath.row];
    if (group.type == TLEmojiTypeOther) {
        //???: 存在冲突：用户选中cellA,再此方法中立马调用方法选中cellB时，所有cell都不会被选中
        dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.01 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
            [self setCurIndexPath:_curIndexPath];
        });
        [self p_eidtMyEmojiButtonDown];
    }else{
        [self setCurIndexPath:indexPath];
    }
}
#pragma mark - Private Methods -
- (void)p_addMasonry{
    [self.addButton mas_makeConstraints:^(MASConstraintMaker *make) {
        make.left.and.top.and.bottom.mas_equalTo(self);
        make.width.mas_equalTo(WIDTH_EMOJIGROUP_CELL);
    }];
    
    [self.sendButton mas_makeConstraints:^(MASConstraintMaker *make) {
        make.top.and.bottom.and.right.mas_equalTo(self);
        make.width.mas_equalTo(WIDTH_SENDBUTTON);
    }];
    
    [self.collectionView mas_makeConstraints:^(MASConstraintMaker *make) {
        make.top.and.bottom.mas_equalTo(self);
        make.left.mas_equalTo(self.addButton.mas_right);
        make.right.mas_equalTo(self.sendButton.mas_left);
    }];
}
- (void)p_eidtMyEmojiButtonDown{
    if (_delegate && [_delegate respondsToSelector:@selector(emojiGroupControlEditMyEmojiButtonDown:)]) {
        [_delegate emojiGroupControlEditMyEmojiButtonDown:self];
    }
}
- (void)drawRect:(CGRect)rect{
    [super drawRect:rect];
    CGContextRef context = UIGraphicsGetCurrentContext();
    CGContextSetLineWidth(context, 0.5);
    CGContextSetStrokeColorWithColor(context, colorGrayLine.CGColor);
    CGContextBeginPath(context);
    CGContextMoveToPoint(context, WIDTH_EMOJIGROUP_CELL, 5);
    CGContextAddLineToPoint(context, WIDTH_EMOJIGROUP_CELL, self.frameHeight - 5);
    CGContextStrokePath(context);
}
#pragma mark - Event Response -
- (void)emojiAddButtonDown{
    if (_delegate && [_delegate respondsToSelector:@selector(emojiGroupControlEditButtonDown:)]) {
        [_delegate emojiGroupControlEditButtonDown:self];
    }
}
- (void)sendButtonDown{
    if (_delegate && [_delegate respondsToSelector:@selector(emojiGroupControlSendButtonDown:)]) {
        [_delegate emojiGroupControlSendButtonDown:self];
    }
}
#pragma mark - Getter -
- (UIButton *)addButton{
    if (_addButton == nil) {
        _addButton = [[UIButton alloc] init];
        [_addButton setImage:[UIImage imageNamed:@"emojiKB_groupControl_add"] forState:UIControlStateNormal];
        [_addButton addTarget:self action:@selector(emojiAddButtonDown) forControlEvents:UIControlEventTouchUpInside];
    }
    return _addButton;
}
- (UICollectionView *)collectionView{
    if (_collectionView == nil) {
        UICollectionViewFlowLayout *layout = [[UICollectionViewFlowLayout alloc] init];
        [layout setScrollDirection:UICollectionViewScrollDirectionHorizontal];
        [layout setMinimumLineSpacing:0];
        _collectionView = [[UICollectionView alloc] initWithFrame:CGRectZero collectionViewLayout:layout];
        [_collectionView setBackgroundColor:[UIColor clearColor]];
        [_collectionView setDataSource:self];
        [_collectionView setDelegate:self];
        [_collectionView setShowsHorizontalScrollIndicator:NO];
        [_collectionView setShowsHorizontalScrollIndicator:NO];
        [_collectionView setScrollsToTop:NO];
    }
    return _collectionView;
}
- (UIButton *)sendButton{
    if (_sendButton == nil) {
        _sendButton = [[UIButton alloc] init];
        [_sendButton.titleLabel setFont:[UIFont systemFontOfSize:15.0f]];
        [_sendButton setTitle:@"  发送" forState:UIControlStateNormal];
        [_sendButton setTitleColor:[UIColor grayColor] forState:UIControlStateNormal];
        [_sendButton setBackgroundColor:[UIColor clearColor]];
        [_sendButton setBackgroundImage:[UIImage imageNamed:@"emojiKB_sendBtn_gray"] forState:UIControlStateNormal];
        [_sendButton setBackgroundImage:[UIImage imageNamed:@"emojiKB_sendBtn_gray"] forState:UIControlStateHighlighted];
        [_sendButton addTarget:self action:@selector(sendButtonDown) forControlEvents:UIControlEventTouchUpInside];
    }
    return _sendButton;
}
@end

File: Freedom/微信/微信/ChatViewController/ChatEmojiKeyboard/TLEmojiBaseCell.swift
////
////  TLEmojiBaseCell.swift
////  Freedom
//import SnapKit
//import Foundation
//class TLEmojiBaseCell: UICollectionViewCell {
//    var emojiItem: TLEmoji = TLEmoji()
//    var bgView = UIImageView()
//    var highlightImage: UIImage?
//    var showHighlightImage = false
//}
//
//class TLEmojiFaceItemCell: TLEmojiBaseCell {
//    private var imageView = UIImageView()
//    override var emojiItem: TLEmoji {
//        didSet {
//            imageView.image = emojiItem.emojiName.isEmpty ? nil : UIImage(named: emojiItem.emojiName)
//        }
//    }
//    override init(frame: CGRect) {
//        super.init(frame: frame)
//        contentView.addSubview(imageView)
//        imageView.snp.makeConstraints { (make) in
//            make.edges.equalToSuperview()
//        }
//    }
//    required init?(coder aDecoder: NSCoder) {
//        fatalError("init(coder:) has not been implemented")
//    }
//}
//
//class TLEmojiImageItemCell: TLEmojiBaseCell {
//    private var imageView = UIImageView()
//    override var emojiItem: TLEmoji {
//        didSet {
//            imageView.image = emojiItem.emojiPath.isEmpty ? nil : UIImage(named: emojiItem.emojiPath)
//        }
//    }
//    override init(frame: CGRect) {
//        super.init(frame: frame)
//        contentView.addSubview(imageView)
//        highlightImage = UIImage(named: "emoji_hl_background")
//        imageView.snp.makeConstraints { (make) in
//            make.edges.equalTo(self.contentView).offset(UIEdgeInsets(top: 2, left: 2, bottom: 2, right: 2) as! ConstraintOffsetTarget)
//        }
//    }
//    required init?(coder aDecoder: NSCoder) {
//        fatalError("init(coder:) has not been implemented")
//    }
//}
//
//class TLEmojiImageTitleItemCell: TLEmojiBaseCell {
//    private var imageView = UIImageView()
//    override var emojiItem: TLEmoji {
//        didSet {
//            imageView.image = emojiItem.emojiPath.isEmpty ? nil : UIImage(named: emojiItem.emojiPath)
//            label.text = emojiItem.emojiName
//        }
//    }
//    private lazy var label: UILabel = {
//        let label = UILabel()
//        label.font = UIFont.systemFont(ofSize: 12.0)
//        label.textColor = UIColor.gray
//        label.textAlignment = .center
//        return label
//    }()
//
//    override init(frame: CGRect) {
//        super.init(frame: frame)
//        highlightImage = UIImage(named: "emoji_hl_background")
//        contentView.addSubview(imageView)
//        contentView.addSubview(label)
////        p_addMasonry()
//    }
//    required init?(coder aDecoder: NSCoder) {
//        fatalError("init(coder:) has not been implemented")
//    }
////    func p_addMasonry() {
////        bgView.mas_remakeConstraints({ make in
////            make.left.and().right().and().top().mas_equalTo(self.contentView)
////            make.height.mas_equalTo(self.bgView.mas_width)
////        })
////        imageView.mas_makeConstraints({ make in
////            make.left.and().top().mas_equalTo(self.contentView).mas_offset(3)
////            make.right.mas_equalTo(self.contentView).mas_offset(-3)
////            make.height.mas_equalTo(self.imageView.mas_width)
////        })
////        label.mas_makeConstraints({ make in
////            make.left.and().right().and().bottom().mas_equalTo(self.contentView)
////        })
////    }
//}
//
//class TLEmojiItemCell: TLEmojiBaseCell {
//    private lazy var label: UILabel = {
//        let label = UILabel()
//        label.font = UIFont.systemFont(ofSize: 25.0)
//        return label
//    }()
//    override var emojiItem: TLEmoji {
//        didSet {
//            label.text = emojiItem.emojiName
//        }
//    }
//    override var showHighlightImage: Bool {
//        didSet {
//            if showHighlightImage {
//                bgView.image = highlightImage
//            } else {
//                bgView.image = nil
//            }
//        }
//    }
//    override init(frame: CGRect) {
//        super.init(frame: frame)
//        bgView.layer.masksToBounds = true
//        bgView.layer.cornerRadius = 5.0
//        contentView.addSubview(label)
//        contentView.addSubview(bgView)
//        bgView.snp.makeConstraints { (make) in
//            make.edges.equalTo(self.contentView)
//        }
//        label.snp.makeConstraints { (make) in
//            make.center.equalToSuperview()
//        }
//    }
//    required init?(coder aDecoder: NSCoder) {
//        fatalError("init(coder:) has not been implemented")
//    }
//}
//
//class TLEmojiGroupCell: UICollectionViewCell {
//    var emojiGroup: TLEmojiGroup {
//        didSet {
//            groupIconView.image = UIImage(named: emojiGroup.groupIconPath)
//        }
//    }
//    private var groupIconView = UIImageView()
//    override init(frame: CGRect) {
//        super.init(frame: frame)
//        backgroundColor = UIColor.clear
//        let selectedView = UIView()
//        selectedView.backgroundColor = UIColor(245.0, 245.0, 247.0, 1.0)
//        selectedBackgroundView = selectedView
//        contentView.addSubview(groupIconView)
//        groupIconView.snp.makeConstraints { (make) in
//            make.center.equalTo(self.contentView)
//            make.width.height.lessThanOrEqualTo(30)
//        }
//    }
//    required init?(coder aDecoder: NSCoder) {
//        fatalError("init(coder:) has not been implemented")
//    }
//    override func draw(_ rect: CGRect) {
//        super.draw(rect)
//        guard let context = UIGraphicsGetCurrentContext() else { return }
//        context.setLineWidth(0.5)
//        context.setStrokeColor(UIColor.gray.cgColor)
//        context.beginPath()
//        context.move(to: CGPoint(x: frame.size.width - 0.5, y: 5))
//        context.addLine(to: CGPoint(x: frame.size.width - 0.5, y: frame.size.height - 5))
//        context.strokePath()
//    }
//}
//protocol TLEmojiGroupControlDelegate: NSObjectProtocol {
//    func emojiGroupControl(_ emojiGroupControl: TLEmojiGroupControl, didSelectedGroup group: TLEmojiGroup)
//    func emojiGroupControlEditButtonDown(_ emojiGroupControl: TLEmojiGroupControl)
//    func emojiGroupControlEditMyEmojiButtonDown(_ emojiGroupControl: TLEmojiGroupControl)
//    func emojiGroupControlSendButtonDown(_ emojiGroupControl: TLEmojiGroupControl)
//}
//
//class TLEmojiGroupControl: UIView,UICollectionViewDataSource, UICollectionViewDelegate, UICollectionViewDelegateFlowLayout {
//    var sendButtonStatus = TLGroupControlSendButtonStatus.none {
//        didSet {
//            if sendButtonStatus == .none {
//                UIView.animate(withDuration: 1, animations: {
//                    self.sendButton.snp.updateConstraints({ (make) in
//                        make.right.equalToSuperview()
//                    })
//                    self.layoutIfNeeded()
//                })
//            }
//            if sendButtonStatus == .none {
//                UIView.animate(withDuration: 1, animations: {
//                    self.sendButton.snp.updateConstraints({ (make) in
//                        make.right.equalToSuperview().offset(60)
//                    })
//                    self.layoutIfNeeded()
//                })
//            } else if sendButtonStatus == .blue {
//                sendButton.setBackgroundImage(UIImage(named: "emojiKB_sendBtn_blue"), for: .normal)
//                sendButton.setBackgroundImage(UIImage(named: "emojiKB_sendBtn_blueHL"), for: .highlighted)
//                sendButton.setTitleColor(UIColor.white, for: .normal)
//            } else if sendButtonStatus == .gray {
//                sendButton.setBackgroundImage(UIImage(named: "emojiKB_sendBtn_gray"), for: .normal)
//                sendButton.setBackgroundImage(UIImage(named: "emojiKB_sendBtn_gray"), for: .highlighted)
//                sendButton.setTitleColor(UIColor.gray, for: .normal)
//            }
//        }
//    }
//    var emojiGroupData: [[TLEmojiGroup]] = [] {
//        didSet {
//            collectionView.reloadData()
//            if emojiGroupData.count > 0 {
//                self.curIndexPath = IndexPath(row: 0, section: 0)
//            }
//        }
//    }
//    weak var delegate: TLEmojiGroupControlDelegate?
//    private var curIndexPath = IndexPath(row: 0, section: 0) {
//        didSet {
//            collectionView.selectItem(at: curIndexPath, animated: false, scrollPosition: [])
//            let group = (emojiGroupData[curIndexPath.section])[curIndexPath.row] as TLEmojiGroup
//            delegate?.emojiGroupControl(self, didSelectedGroup: group)
//        }
//    }
//    private lazy var addButton : UIButton = {
//        let addButton = UIButton()
//        addButton.setImage(UIImage(named: "emojiKB_groupControl_add"), for: .normal)
//        addButton.addTarget(self, action: #selector(self.emojiAddButtonDown), for: .touchUpInside)
//        return addButton
//    }()
//    private lazy var collectionView: UICollectionView =  {
//        let layout = UICollectionViewFlowLayout()
//        layout.scrollDirection = .horizontal
//        layout.minimumLineSpacing = 0
//        let collectionView = UICollectionView(frame: CGRect.zero, collectionViewLayout: layout)
//        collectionView.backgroundColor = UIColor.clear
//        collectionView.dataSource = self
//        collectionView.delegate = self
//        collectionView.showsHorizontalScrollIndicator = false
//        collectionView.showsHorizontalScrollIndicator = false
//        collectionView.scrollsToTop = false
//        return collectionView
//    }()
//
//    private lazy var sendButton:UIButton = {
//        let sendButton = UIButton()
//        sendButton.titleLabel?.font = UIFont.systemFont(ofSize: 15.0)
//        sendButton.setTitle("发送", for: .normal)
//        sendButton.setTitleColor(UIColor.gray, for: .normal)
//        sendButton.backgroundColor = UIColor.clear
//        sendButton.setBackgroundImage(UIImage(named: "emojiKB_sendBtn_gray"), for: .normal)
//        sendButton.setBackgroundImage(UIImage(named: "emojiKB_sendBtn_gray"), for: .highlighted)
//        sendButton.addTarget(self, action: #selector(self.sendButtonDown), for: .touchUpInside)
//        return sendButton
//    }()
//    override init(frame: CGRect) {
//        super.init(frame: frame)
//        backgroundColor = UIColor.white
//        addSubview(addButton)
//        addSubview(collectionView)
//        addSubview(sendButton)
////        p_addMasonry()
//        collectionView.register(TLEmojiGroupCell.self, forCellWithReuseIdentifier: "TLEmojiGroupCell")
//    }
//
//    required init?(coder aDecoder: NSCoder) {
//        fatalError("init(coder:) has not been implemented")
//    }
//
//    func numberOfSections(in collectionView: UICollectionView) -> Int {
//        return emojiGroupData.count
//    }
//
//    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
//        return (emojiGroupData[section] as [Any]).count
//    }
//
//    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
//        let cell = collectionView.dequeueReusableCell(withReuseIdentifier: "TLEmojiGroupCell", for: indexPath) as! TLEmojiGroupCell
//        let group = emojiGroupData[indexPath.section][indexPath.row] as TLEmojiGroup
//        cell.emojiGroup = group
//        return cell
//    }
//
//    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, sizeForItemAt indexPath: IndexPath) -> CGSize {
//        return CGSize(width: 46, height: frame.size.height)
//    }
//    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, referenceSizeForFooterInSection section: Int) -> CGSize {
//        if section == emojiGroupData.count - 1 {
//            return CGSize(width: 46 * 2, height: frame.size.height)
//        }
//        return CGSize.zero
//    }
//    func collectionView(_ collectionView: UICollectionView, didSelectItemAt indexPath: IndexPath) {
//        let group = emojiGroupData[indexPath.section][indexPath.row] as TLEmojiGroup
//        if group.type == .other {
//            //: 存在冲突：用户选中cellA,再此方法中立马调用方法选中cellB时，所有cell都不会被选中
//            DispatchQueue.main.asyncAfter(deadline: DispatchTime.now() + Double(Int64(0.01 * Double(NSEC_PER_SEC))) / Double(NSEC_PER_SEC), execute: {
//                self.curIndexPath = self.curIndexPath
//            })
//            p_eidtMyEmojiButtonDown()
//        } else {
//            self.curIndexPath = indexPath
//        }
//    }
////    func p_addMasonry() {
////        addButton.mas_makeConstraints({ make in
////            make.left.and().top().and().bottom().mas_equalTo(self)
////            make.width.mas_equalTo(46)
////        })
////        sendButton.mas_makeConstraints({ make in
////            make.top.and().bottom().and().right().mas_equalTo(self)
////            make.width.mas_equalTo(60)
////        })
////        collectionView.mas_makeConstraints({ make in
////            make.top.and().bottom().mas_equalTo(self)
////            make.left.mas_equalTo(self.addButton.mas_right)
////            make.right.mas_equalTo(self.sendButton.mas_left)
////        })
////    }
//    func p_eidtMyEmojiButtonDown(){
//        delegate?.emojiGroupControlEditMyEmojiButtonDown(self)
//    }
//
//    override func draw(_ rect: CGRect) {
//        super.draw(rect)
//        guard let context = UIGraphicsGetCurrentContext() else { return }
//        context.setLineWidth(0.5)
//        context.setStrokeColor(UIColor.gray.cgColor)
//        context.beginPath()
//        context.move(to: CGPoint(x: 46, y: 5))
//        context.addLine(to: CGPoint(x: 46, y: frame.size.height - 5))
//        context.strokePath()
//    }
//    @objc func emojiAddButtonDown() {
//        delegate?.emojiGroupControlEditButtonDown(self)
//    }
//    @objc func sendButtonDown() {
//        delegate?.emojiGroupControlSendButtonDown(self)
//    }
//}
